### PR TITLE
Neox 2.x to master-2.x

### PR DESCRIPTION
**Pull request by superboyiii**
Neox 2.x to master-2.x

### CODE REVIEW

File: ImportBlocks/BlockImporter.cs [hunk 1]
```diff
@@ -34,7 +36,7 @@ private static IEnumerable<Block> GetBlocks(Stream stream, bool read_start = fal
             {
                 uint start = read_start ? r.ReadUInt32() : 0;
                 uint count = r.ReadUInt32();
-                uint end = start + count - 1;
+                uint end = checked(start + count) - 1;
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:39]:
```suggestion
                uint end = checked(start + count - 1);
```

0+0-1 ?

File: ImportBlocks/BlockImporter.cs [hunk 2]
```diff
@@ -49,6 +51,53 @@ private static IEnumerable<Block> GetBlocks(Stream stream, bool read_start = fal
             }
         }
 
+        private static IEnumerable<StateRoot> GetStateRoots(Stream stream)
+        {
+            using (BinaryReader r = new BinaryReader(stream))
+            {
+                uint start = r.ReadUInt32();
+                uint count = r.ReadUInt32();
+                uint end = checked(start + count) - 1;
```
Comment (shargon) [inline ImportBlocks/BlockImporter.cs:60]:
```suggestion
                uint end = checked(start + count - 1);
```

File: ImportBlocks/BlockImporter.cs (hunk 1 above)
Comment (KickSeason) [reply ImportBlocks/BlockImporter.cs:39]: done https://github.com/neo-project/neo-modules/pull/323

File: ImportBlocks/BlockImporter.cs (hunk 2 above)
Comment (KickSeason) [reply ImportBlocks/BlockImporter.cs:60]: done https://github.com/neo-project/neo-modules/pull/323

File: RpcNep5Tracker/RpcNep5Tracker.cs [hunk 3]
```diff
@@ -192,7 +192,7 @@ public void OnPersist(Snapshot snapshot, IReadOnlyList<Blockchain.ApplicationExe
                     script = sb.ToArray();
                 }
 
-                using (ApplicationEngine engine = ApplicationEngine.Run(script, snapshot, extraGAS: maxGas))
+                using (ApplicationEngine engine = ApplicationEngine.Run(script, snapshot.Clone(), extraGAS: maxGas))
```
Comment (vncoelho) [inline RpcNep5Tracker/RpcNep5Tracker.cs:195]: This is for safety reasons, right?

Comment (superboyiii) [reply RpcNep5Tracker/RpcNep5Tracker.cs:195]: > This is for safety reasons, right?

Yes, beacuse some bugs were caused if not clone here.

**Comment by superboyiii** [comment]
> @superboyiii, this is mostly for importing/exporting, right?

Yes, and support root import/export now.

**Comment by superboyiii** [comment]
@shargon Could you review it again? Now it's fine.

File: ImportBlocks/ImportBlocks.cs [hunk 4]
```diff
@@ -23,31 +23,91 @@ private bool OnExport(string[] args)
...
+                if (start <= end)
+                    fs.Write(BitConverter.GetBytes(count), 0, sizeof(uint));
+                fs.Seek(0, SeekOrigin.End);
+                for (uint i = start; i <= end; i++)
+                {
+                    StateRootState state = Blockchain.Singleton.Store.GetStateRoots().TryGet(i);
+                    byte[] array = state.StateRoot.ToArray();
+                    fs.Write(BitConverter.GetBytes(array.Length), 0, sizeof(int));
+                    fs.Write(array, 0, array.Length);
+                    Console.SetCursorPosition(0, Console.CursorTop);
```
Comment (shargon) [inline ImportBlocks/ImportBlocks.cs:105]: We should use the progress bar https://github.com/neo-project/neo-node/blob/master/neo-cli/CLI/ConsolePercent.cs

Comment (superboyiii) [reply ImportBlocks/ImportBlocks.cs:105]: > We should use the progress bar https://github.com/neo-project/neo-node/blob/master/neo-cli/CLI/ConsolePercent.cs

Maybe we should merge this first and pull another request for progress bar.

**Review by shargon — APPROVED** [review]

**Pull request merged**